theano/sandbox/cuda/opt.py

```diff
@@ -440,6 +441,24 @@ def local_gpu_dot_to_dot22(node):
     return False
 
 
+@local_optimizer([None])
+def local_assert_no_cpu_op(node):
+    if not isinstance(node.op, GpuOp) and all([var.owner and isinstance(var.owner.op,
+        HostFromGpu) for var in node.inputs]) and all([var.owner and
+            isinstance(var.owner.op, GpuFromHost) for var in node.outputs]):
+            if config.assert_no_cpu_op == "warn":
+                _logger.warning(("CPU op %s is detected in the computational"
+                                 " graph") % node)
+            elif config.assert_no_cpu_op == "raise":
+                raise RuntimeError("The op %s is on CPU." % node)
+            elif config.assert_no_cpu_op == "pdb":
+                pdb.set_trace()
+    return None
+
+if config.assert_no_cpu_op != "ignore":
+    register_opt()(local_assert_no_cpu_op)
```

Comment:
This opt should be registered later, not during the optimization that move the computation to the GPU.

I think this should be done in the speciliaze_device equilibrium.

To register it there:
theano.compile.optdb['specialize_device'].register(...